File: frontend/src/api/bootstrapPlan.ts
import type { DeploymentConfigurationPreview } from "./deploymentConfiguration";
import type { CurrentIdentity } from "./identity";
import type { ReleasePreflight } from "./releasePreflight";
import { apiFetch } from "./client";

export type BootstrapPlan = {
  plan_id: string;
  schema_version: string;
  release_id: string;
  profile: string;
  organization_id: string;
  environment_id: string;
  site_id: string;
  state: "ready" | "blocked";
  plan_digest: string;
  resume_key: string;
  phases: Array<{
    phase_id: string;
    sequence: number;
    title: string;
    dependencies: string[];
    state: "ready" | "blocked";
    resumable: boolean;
    input_references: string[];
    stop_guidance: string;
  }>;
  generated_at: string;
  correlation_id: string;
  mutation_authorized: false;
  execution_authorized: false;
};

type BootstrapPlanResponse = { data: BootstrapPlan };

function isPlanResponse(value: unknown): value is BootstrapPlanResponse {
  if (typeof value !== "object" || value === null || !("data" in value)) return false;
  const data = value.data;
  return typeof data === "object" && data !== null && "phases" in data && Array.isArray(data.phases) && "plan_digest" in data && typeof data.plan_digest === "string" && "state" in data && (data.state === "ready" || data.state === "blocked");
}

export async function getBootstrapPlan(
  preflight: ReleasePreflight,
  configuration: DeploymentConfigurationPreview,
  scope: CurrentIdentity["scope"],
): Promise<BootstrapPlanResponse | null> {
  const response = await apiFetch("/api/v1/platform/bootstrap-plan", {
    method: "POST",
    headers: { Accept: "application/json", "Content-Type": "application/json" },
    body: JSON.stringify({
      schema_version: "atlas.bootstrap-plan-request.v1",
      release_id: preflight.release_id,
      profile: preflight.profile,
      organization_id: scope.organization_id,
      environment_id: scope.environment_id,
      site_id: scope.site_id,
      preflight_report_id: preflight.report_id,
      manifest_digest: preflight.manifest_digest,
      preflight_state: preflight.state,
      configuration_preview_id: configuration.preview_id,
      configuration_digest: configuration.configuration_digest,
      configuration_state: configuration.state,
    }),
  });
  if (response.status === 403) return null;
  if (!response.ok) throw new Error(`Bootstrap plan failed with ${response.status}`);
  const payload: unknown = await response.json();
  return isPlanResponse(payload) ? payload : null;
}

File: backend/src/atlas/modules/platform/domain/bootstrap_plan.py
from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum

from atlas.modules.identity.domain.models import validate_stable_identifier
from atlas.modules.platform.domain.release_preflight import DeploymentProfile

DIGEST_PATTERN = re.compile(r"^[a-f0-9]{64}$")


class BootstrapPlanState(StrEnum):
    READY = "ready"
    BLOCKED = "blocked"


class BootstrapPhaseState(StrEnum):
    READY = "ready"
    BLOCKED = "blocked"


@dataclass(frozen=True, slots=True)
class BootstrapPlanRequest:
    schema_version: str
    release_id: str
    profile: DeploymentProfile
    organization_id: str
    environment_id: str
    site_id: str
    preflight_report_id: str
    manifest_digest: str
    preflight_state: str
    configuration_preview_id: str
    configuration_digest: str
    configuration_state: str

    def __post_init__(self) -> None:
        if self.schema_version != "atlas.bootstrap-plan-request.v1":
            raise ValueError("bootstrap plan request schema is unsupported")
        for value, label in (
            (self.release_id, "release_id"),
            (self.organization_id, "organization_id"),
            (self.environment_id, "environment_id"),
            (self.site_id, "site_id"),
            (self.preflight_report_id, "preflight_report_id"),
            (self.configuration_preview_id, "configuration_preview_id"),
        ):
            validate_stable_identifier(value, label)
        if not DIGEST_PATTERN.fullmatch(self.manifest_digest) or not DIGEST_PATTERN.fullmatch(
            self.configuration_digest
        ):
            raise ValueError("bootstrap plan input digest is invalid")
        if self.preflight_state not in {"passed", "failed", "warning", "unchecked"}:
            raise ValueError("preflight state is invalid")
        if self.configuration_state not in {"passed", "failed"}:
            raise ValueError("configuration state is invalid")


@dataclass(frozen=True, slots=True)
class BootstrapPhase:
    phase_id: str
    sequence: int
    title: str
    dependencies: tuple[str, ...]
    state: BootstrapPhaseState
    resumable: bool
    input_references: tuple[str, ...]
    stop_guidance: str


@dataclass(frozen=True, slots=True)
class BootstrapPlan:
    plan_id: str
    schema_version: str
    release_id: str
    profile: DeploymentProfile
    organization_id: str
    environment_id: str
    site_id: str
    state: BootstrapPlanState
    plan_digest: str
    resume_key: str
    phases: tuple[BootstrapPhase, ...]
    generated_at: datetime
    correlation_id: str
    mutation_authorized: bool = False
    execution_authorized: bool = False
